### ISSUE TITLE
Ram usage for cache

### ISSUE THREAD

**Issue by Homer-Sim** [syncthing/syncthing#4175]
Hello,

I am running syncthing server on Ubuntu 16.04 containered. It is now setup for 2 weeks and syncing is working fine. I recognized a lot of ram usage (4GB) for caching. With newly installed netdata I figured out it is syncthing using that much.

Is there a setting option which can be set to limit ram cache size?

Thanks

**Comment by calmh** [comment]
How do you see / measure this usage?

**Comment by Ferroin** [comment]
So, there's this thing that the _OS_ (not Syncthing) has called a filesystem cache (well, technically it's more correctly a VFS cache or page-cache depending on who you ask).  Unless you hooked up to Syncthing with a debugger and between that and code inspection figured out exactly how it's using every last bit of memory it allocates (and seeing serious bugs in Syncthing, because there is no reason Syncthing itself should be allocating 4G), this is what's using that 4G of RAM.

When a block of a file gets read off disk, it has to be copied into memory, this is very slow compared to just accessing memory directly, so the OS will keep a copy of that block around in RAM (if there's room) to save time if the block needs to be read again later (because then it can just copy the data from the cache directly to the program that needs it).   In the event that memory used by the cache is needed elsewhere, the least recently used entry in the cache is removed from memory, and that space gets reused.

Put simply, this is the OS making efficient use of RAM that would otherwise be unused.  Netdata is probably reporting Syncthing as the source because Syncthing made the read and write calls that resulted in that data being in the cache (file sharing apps tend to fill the cache pretty quickly on startup because of how they work).

**Comment by calmh** [comment]
Yes. I just want to make sure we are talking about fs cache, in which case this is indeed functions-as-designed.

**Comment by Ferroin** [comment]
If he's seeing it through netdata though, it pretty much has to be FS cache, since netdata doesn't do ptrace and code analysis, which would be required to determine that allocations internal to syncthing are being used for caching.

**Comment by Homer-Sim** [comment]
Hello and thanks so far for your answers. In netdata my single containers are listed and only in syncthing container the cache appears. This made me think it could be an syncthing issue.

**Comment by AudriusButkevicius** [comment]
I mean, is it an actual issue or you just saw some number that you potentially do not understand and decided it's an issue.

If the usage is high in `top` commandline utility in the RES column, then it's an issue, otherwise we are talking either about kernel page caches that do not matter as they get dropped as need for memory comes or about virtual memory space which doesn't matter as it's not real memory.

**Comment by Ferroin** [comment]
Expanding on @AudriusButkevicius comment, high filesystem cache usage is actually expected behavior for applications like Syncthing, it has to index all the files in the synced directories, which in turn means that it has to read all the data in all those files.  You should see similar behavior from most other sync software, including things like Dropbox and BTSync.

That said, unless you are seeing performance issues, the only memory usage aspects you should be worrying about most of the time are the resident-set-size ('RES' in `top` and friends, 'used' in netdata and most other data collection software), which is how much physical RAM is being used directly by the application, and swap utilization, as those are what dictate how much space you can use before the system runs out of memory.

**Comment by calmh** [comment]
Closing for lack of info / reproducibility
